Custom MCDU Font

tyler58546 opened the pull request:
## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
* Added custom MCDU font

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->
Before:
![image](https://user-images.githubusercontent.com/16512508/96042477-4f79ee80-0e22-11eb-8ad3-c6510fbf103d.png)
After:
![image](https://user-images.githubusercontent.com/16512508/96042598-7df7c980-0e22-11eb-8fdf-23a93c45de5a.png)
![image](https://user-images.githubusercontent.com/16512508/96043024-3160be00-0e23-11eb-9dcc-92b970b48a95.png)
![image](https://user-images.githubusercontent.com/16512508/96043053-3e7dad00-0e23-11eb-90c2-0df06906e01e.png)


## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->
![image](https://user-images.githubusercontent.com/16512508/96042737-b8616680-0e22-11eb-85fc-a7e69cdf32fa.png)
![image](https://user-images.githubusercontent.com/16512508/96042858-e9da3200-0e22-11eb-8779-3191a69ffd53.png)

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build.py will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, slick on the **Artifacts** drop down and click the **A32NX** link

Externoak commented:
Is char ⇅ (21C5) working in the IRS init page with this font? @wpine215

Lucky38i commented:
you could always replace my `_`s with a special identifier like `{space}` just replace the `_` on https://github.com/flybywiresim/a32nx/blob/fde6073fafef6720bafdfe28cc5cdb04918ea47c/A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js#L411 and https://github.com/flybywiresim/a32nx/blob/fde6073fafef6720bafdfe28cc5cdb04918ea47c/A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js#L415

tyler58546 commented:
@Lucky38i I fixed the spacing issues and adjusted the formatting system so it works on all lines of all pages now.

Lucky38i commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:
```diff
@@ -44,20 +44,20 @@ class CDUIRSInit {
+            GPSPosAlign = ["--d--.--", "--d--.--", " "];
+            originAirportString = ['{IrsInitFont}' + originAirportLat['deg'] + 'd{IrsInitFontEnd}' + originAirportLat['min'] + '.' + originAirportLat['sec'] + '[s-text]{IrsInitFont}' + originAirportLat['dir'] + "{IrsInitFontEnd} [color]blue", '{IrsInitFont}' + originAirportLon['deg'] + 'd{IrsInitFontEnd}' + originAirportLon['min'] + '.' + originAirportLon['sec'] + '[s-text]{IrsInitFont}' + origin ...
             if (SimVar.GetSimVarValue("L:A320_Neo_ADIRS_KNOB_1", "Enum") || SimVar.GetSimVarValue("L:A320_Neo_ADIRS_KNOB_2", "Enum") || SimVar.GetSimVarValue("L:A320_Neo_ADIRS_KNOB_3", "Enum")) {
-                GPSPosAlign = ['{IrsInitFont}' + currentGPSLat['deg'] + '°{IrsInitFontEnd}' + currentGPSLat['min'] + '.' + Math.ceil(Number(currentGPSLat['sec'] / 100)) + '[s-text]{IrsInitFont}' + currentGPSLat['dir'] + "{IrsInitFontEnd} [color]green", '{IrsInitFont}' + currentGPSLon['deg'] + '°{IrsInitFontEnd}' + currentGPSLon['min'] + '.' + Math.ceil(Number(currentGPSLon['sec'] / 100)) + '[s-te ...
+                GPSPosAlign = ['{IrsInitFont}' + currentGPSLat['deg'] + 'd{IrsInitFontEnd}' + currentGPSLat['min'] + '.' + Math.ceil(Number(currentGPSLat['sec'] / 100)) + '[s-text]{IrsInitFont}' + currentGPSLat['dir'] + "{IrsInitFontEnd} [color]green", '{IrsInitFont}' + currentGPSLon['deg'] + 'd{IrsInitFontEnd}' + currentGPSLon['min'] + '.' + Math.ceil(Number(currentGPSLon['sec'] / 100)) + '[s-te ...
                 alignType = "GPS";
             }
         }
 
-        let IRSAlignOnPos = "{IrsInitFont}" + currentGPSLat['deg'] + '°{IrsInitFontEnd}' + currentGPSLat['min'] + '.' + Math.ceil(Number(currentGPSLat['sec'] / 100)) + '[s-text]{IrsInitFont}' + currentGPSLat['dir'] + '/{IrsInitFontEnd}{IrsInitFont}' + currentGPSLon['deg'] + '°{IrsInitFontEnd}' + currentGPSLon['min'] + '.' + Math.ceil(Number(currentGPSLon['sec'] / 100)) + '{IrsInitFont}' + current ...
+        let IRSAlignOnPos = "{IrsInitFont}" + currentGPSLat['deg'] + 'd{IrsInitFontEnd}' + currentGPSLat['min'] + '.' + Math.ceil(Number(currentGPSLat['sec'] / 100)) + '[s-text]{IrsInitFont}' + currentGPSLat['dir'] + '/{IrsInitFontEnd}{IrsInitFont}' + currentGPSLon['deg'] + 'd{IrsInitFontEnd}' + currentGPSLon['min'] + '.' + Math.ceil(Number(currentGPSLon['sec'] / 100)) + '{IrsInitFont}' + current ...
         if (SimVar.GetSimVarValue("L:A32XN_Neo_ADIRS_ALIGN_TYPE_REF", "Enum") === 1) {
             alignMsg = "";
             alignType = "REF";
             if (checkAligned !== 2) {
-                IRSAlignOnPos = "{IrsInitFont}" + originAirportLat['deg'] + '°{IrsInitFontEnd}' + originAirportLat['min'] + '.' + originAirportLat['sec'] + '[s-text]{IrsInitFont}' + originAirportLat['dir'] + '/{IrsInitFontEnd}{IrsInitFont}' + originAirportLon['deg'] + '°{IrsInitFontEnd}' + originAirportLon['min'] + '.' + originAirportLon['sec'] + '{IrsInitFont}' + originAirportLon['dir'] + '{IrsI ...
+                IRSAlignOnPos = "{IrsInitFont}" + originAirportLat['deg'] + 'd{IrsInitFontEnd}' + originAirportLat['min'] + '.' + originAirportLat['sec'] + '[s-text]{IrsInitFont}' + originAirportLat['dir'] + '/{IrsInitFontEnd}{IrsInitFont}' + originAirportLon['deg'] + 'd{IrsInitFontEnd}' + originAirportLon['min'] + '.' + originAirportLon['sec'] + '{IrsInitFont}' + originAirportLon['dir'] + '{IrsI ...
```
Based on the changes you recently made it may be useful to rearrange this too to make use of your super method of inserting span text. For example {IrsInit}[content]{end} would make for tider cod

Lucky38i reviewed (changes requested):
In addition to my requested changes, it might be useful to create a file or edit the contributors readme stating what `{sp}` `{end}` etc. etc. do incase they desire to implement something similiar and instead can use our pre-existing solution

tyler58546 commented:
I have renamed `{IrsInitFont}` to `{large}` and fixed some css issues with the IRS INIT page.

lhoenig commented:
Nice work! Don't forget the upper case delta for the OVFY button! ΔΔΔ

lhoenig commented:
I really love the font, unfortunately the beautiful pixel art is messed up pretty bad in the current state by some anti aliasing or other font processing.. look at the slash for example :(
![aa_bad](https://user-images.githubusercontent.com/4353030/96168822-9c90ba00-0f21-11eb-982f-818941c1ec47.PNG)

Lucky38i commented:
> I have renamed `{IrsInitFont}` to `{large}` and fixed some css issues with the IRS INIT page.

`IrsInitFont` uses `irs-text` as a span class that @Externoak created, I'd suggest you'd stick with that span class and use a more appropriate parameter like `IrsFont`

lhoenig commented:
Another problem, I can't see the degrees character on the Closest Airports page, for example.

tyler58546 commented:
> 
> 
> > I have renamed `{IrsInitFont}` to `{large}` and fixed some css issues with the IRS INIT page.
> 
> `IrsInitFont` uses `irs-text` as a span class that @Externoak created, I'd suggest you'd stick with that span class and use a more appropriate parameter like `IrsFont`

The `irs-text` span just sets the font size to its original large size when it is already inside a `s-text` div. I renamed this class so its easier to understand what it does.


> 
> 
> Another problem, I can't see the degrees character on the Closest Airports page, for example.

This has been fixed.

Carter633 reviewed (approved):
Name: British Avgeek YT -- Carter633 #1436
Date of testing: Current date 17/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1342

Testing technique: Full flight from Cold and Dark
Steps you've made to test the Pull Request:
1. Spawned Cold & Dark
2. Started avionics - inputted all data into MCDU
3. Reviewed font throughout all input and across all pages
4. Completed full flight

Overall Rating: Good
Conclusions of the testing: No issues found, looks nice and crisp for me. LGTM!

St54Kevin commented:
Cant test. Reinstalled latest dev and try to copy this over. The result is this:
![image](https://user-images.githubusercontent.com/448627/96498657-a0d41480-124c-11eb-9354-b64ce782c674.png)

Benjozork commented:
@St54Kevin please go to discord for installation assistance.

lhoenig commented:
Here's a picture of the OVFY triangle glyph that appears on the scratchpad when that button is pressed! Big thanks to our member who was able to produce it! Keep in mind it is from a Thales unit though. The fonts should be very similar though.
![20201019_142131_2](https://user-images.githubusercontent.com/4353030/96636165-2f17cb80-131d-11eb-973c-d3e15b0a6071.jpg)

Carter633 commented:
Name: British Avgeek YT -- Carter633 #1436
Date of testing: Current date 23/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1342

Testing technique: Full flight from Cold and Dark
Steps you've made to test the Pull Request: Repeated same flight and steps as first test to check for differences

1.Spawned Cold & Dark
2.Started avionics - inputted all data into MCDU
3.Reviewed font throughout all input and across all pages
4.Completed full flight

Overall Rating: Good
Conclusions of the testing: No issues found, looks nice - thanks for the great work!

armankir commented:
Name: Arman#8145
Date of testing: Current date 23/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1342 

Testing technique: Flight setup on ground, went through _EVERY_ page to see any issues.

Steps you've made to test the Pull Request:

Spawn C&D
Make a full flight plan on Simbreif with no airways
Run full flight setup in MCDU, then look at every remaining page I can and look for text issues
Verify with a second run scenario.


Overall Rating: Good
Conclusions of the testing: Great work on this font! Not much to say besides it looks much better than the default, keep up the good work!

TenPatrol commented:
Name: Tenpatrol
Date of testing: 27/10/2020 22:00Z
Version of the sim: 1.9.5.0
PR Tested: #1342

Testing technique: Full flightg.
Steps you've made to test the Pull Request:
spawn cold & dark
MCDU preparation
full flight

Overall rating: 7/10
Conclusions of the testing: I appreciate your hard work but "slash" and "s" look horrible.
Sorry for being honest.

I though you may be intersested https://github.com/polarsys/b612

TenPatrol commented:
Name: Tenpatrol
Date of testing: 27/10/2020 22:00Z
Version of the sim: 1.9.5.0
PR Tested: #1342

Testing technique: Full flightg.
Steps you've made to test the Pull Request:
spawn cold & dark
MCDU preparation
full flight

Overall rating: 7/10
Conclusions of the testing: I appreciate your hard work but "slash", "s" look horrible.

![MCDU_1](https://user-images.githubusercontent.com/71746362/97246032-a86f5c80-17fc-11eb-8f3c-c3488f80157d.jpg)

![mcdu_2](https://user-images.githubusercontent.com/71746362/97249554-65b18280-1804-11eb-96d4-67857373aeef.jpg)

Sorry for being honest.

I think you may be interested https://github.com/polarsys/b612

ChruutvoLuzi commented:
Name: ChruutvoLuzi | SD-PB.ch#8902
Date of testing: 28.10.2020
Version of the sim: 1.9.5.0
PR Tested: #1342 
Object of test: new MCDU Font

Testing technique: Started cold and Dark and started the a/c
Steps you've made to test the Pull Request:

 Spawned at Gate A44 in Zurich
 Turned on batteries
 checked through the mcdu

Flight notes
Startup: Nothing extraordinary
Taxi/Takeoff: n/a 
Cruise: n/a
Descent/Landing: n/a
Shutdown: n/a

Issues (If Any): Font Insertion works without any issues, however idk if its a sim issue or from the font but some pixel seem bigger and off (see picture below)
Overall Rating: A very nice PR to improve the overall system look.
Conclusions of the testing: if the issue with the pixel is fixable, i would to that prior to merging because it looks quite bad to be honest. Otherwise, the font (as it is shown in the PR) looks very nice

![PR1342](https://user-images.githubusercontent.com/70195216/97487439-2a769700-195d-11eb-88fd-4e2bd83e91a2.png)

beheh commented:
From the last screenshot it looks like the equal sign (=) has gone missing, as used on the PERF pages:

![grafik](https://user-images.githubusercontent.com/929769/98523984-e2b61080-2276-11eb-8493-685dce2d9573.png)

stevex77 commented:
I was just thinking about the font, finished with following conclusion:

I think that using raster font like this will always produce bad output. The reason seems to be, when investigating deeper, that on the real MCDU font pixels are 1:1 (or in whole multiplications) mapped to physical pixels of MCDU display. It's not possible todo this on our MCDU viewport / within virtual cockpit on our computer displays.

Therefore I think that re-do the vector based fond will be better way. This type of font will be always rendered "clean" and "crisp" despite of zoom level and viewing angle of MCDU display.

Anyway, within the latest ASOBO update 1.10.11.0 they reworked temporal antialiasing and sharpening, so now everyting in the cockpit is much sharper and more readable, even zooming out.

beheh commented:
@stevex77 So I understand this is a bitmap font right now in the PR - are you saying you believe with a vector font (containing a path for the pixelated outline) you expect the visual fidelity to be much higher? Or are you saying you don't believe a pixelated font will look good regardless due to text antialiasing?

stevex77 commented:
@beheh exactly. Bitmap / raster fonts are particulary good when its pixels can be exactly mapped to physical display pixels or scaled in exact whole multiplications (1, 2, 3, ...). The will not look good when used on viewports that will be freely scaled and 3D rotated, like our DUs in virtual cockpit. Vector fonts will be handled better in that cases.

Here is an example of raster font mapped 1:1 to physical pixels. Note that these fonts are not antialiased, their pixels are presented "as they are":
![Console](https://user-images.githubusercontent.com/51388935/98855622-9debce80-245c-11eb-8702-59a4af64fad9.png)

Vector fonts are designed to be freely tranformed (rotated, scaled in and out, deformed, ...) and they look better when antialiased.

When try to scale bitmap fonts, let's say 12px of height to 15 physical displays on screen, it will end up with some (three pixels in this case, somehow "randomly" choosen) pixels will be there twice and the rest of pixels just once. It will look worse when applying some anisotropic filtering when viewed from angle (typical scenario within VC).

Inexorable95 commented:
Hey, this is a somewhat older PR, but I think I may have a simple fix for the rendering issues. I'm not 100% sure as I've not been able to test it just yet, but it's an idea (I'll test it tonight and come back to you guys)
We don't actually need to use vectors, because what we want here is to see individual pixels in a crisp manner, a vector would work for this, but it would cause overhead compared to a simple bitmap because vectors are being rendered, whereas pixels are just being displayed as is, and it would require someone to do the SVG while we have the bitmap available now.

What we need to do instead, is to tell the renderer to render individual pixels as they are (square) instead of trying to smooth them out like most web renders do (subpixel rendering), we just need a few css properties for that:

I've been told the mcdu runs on safari, but I'm not so sure, but the properties should be as followed:


	image-rendering: -moz-crisp-edges;          /* Firefox                        */

	image-rendering: optimizeSpeed;             /* STOP SMOOTHING, GIVE ME SPEED  */

	image-rendering: -o-crisp-edges;            /* Opera                          */

	image-rendering: -webkit-optimize-contrast; /* Chrome (and eventually Safari) */

	image-rendering: pixelated; /* Chrome */

	image-rendering: optimize-contrast;         /* CSS3 Proposed                  */

	-ms-interpolation-mode: nearest-neighbor;   /* IE8+                           */

For reference, here is the minecraft 16x16 dirt texture upscaled to 500px x 500px in Chrome:
![minecraft_dirt](https://user-images.githubusercontent.com/76154362/102466713-af258d80-404f-11eb-80ba-8bbcd55d326f.PNG)
Now here's the same texture being upscaled with the CSS properties I mentionned earlier
![minecraft_dirt2](https://user-images.githubusercontent.com/76154362/102466778-c49ab780-404f-11eb-911d-1b3c6db3a75c.PNG)
This is an extreme example, but it clearly shows that the subpixel rendering done by the browser is good for somewhat large pictures, but horrible when it comes to displaying individual pixels, when it is removed, we can upscale the pixels and see them as crisp square objects no matter how much we zoom in

Inexorable95 commented:
Ok, it seems the problem isn't that the pixels are being blurred out, the font is being displayed as a vector font, but the squares are being smoothed out, either by a css property (I haven't been able to track it) or by the game engine, either way, I have given up trying to display single pixels accurately (as squares) on the CDU, besides, the font isn't the correct one for the Honeywell CDU.

The original contributor seems to have given up on this issue, and so have I

pareil6 commented:
Just throwing one more thought into the mix (as someone who has no idea what's going on here) - is this potentially an artifact of the post-processing that's being done by the game? See [this thread](https://www.avsim.com/forums/topic/582923-turn-off-the-forced-sharpening-filter-for-a-cleaner-image/) for what I mean - wondering if the sharpening algorithm is potentially doing this or something?
